driver-core/src/main/com/mongodb/internal/operation/Operations.java

```diff
@@ -0,0 +1,511 @@
+import com.mongodb.client.model.FindOneAndDeleteOptions;
+import com.mongodb.client.model.FindOneAndReplaceOptions;
+import com.mongodb.client.model.FindOneAndUpdateOptions;
+import com.mongodb.client.model.FindOptions;
+import com.mongodb.client.model.IndexModel;
+import com.mongodb.client.model.InsertManyOptions;
+import com.mongodb.client.model.InsertOneModel;
+import com.mongodb.client.model.InsertOneOptions;
+import com.mongodb.client.model.MapReduceAction;
+import com.mongodb.client.model.RenameCollectionOptions;
+import com.mongodb.client.model.ReplaceOneModel;
+import com.mongodb.client.model.ReturnDocument;
+import com.mongodb.client.model.UpdateManyModel;
+import com.mongodb.client.model.UpdateOneModel;
+import com.mongodb.client.model.UpdateOptions;
+import com.mongodb.client.model.WriteModel;
+import com.mongodb.operation.AggregateOperation;
+import com.mongodb.operation.AggregateToCollectionOperation;
+import com.mongodb.operation.CountOperation;
+import com.mongodb.operation.CreateIndexesOperation;
+import com.mongodb.operation.DistinctOperation;
+import com.mongodb.operation.DropCollectionOperation;
+import com.mongodb.operation.DropIndexOperation;
+import com.mongodb.operation.FindAndDeleteOperation;
+import com.mongodb.operation.FindAndReplaceOperation;
+import com.mongodb.operation.FindAndUpdateOperation;
+import com.mongodb.operation.FindOperation;
+import com.mongodb.operation.ListCollectionsOperation;
+import com.mongodb.operation.ListDatabasesOperation;
+import com.mongodb.operation.ListIndexesOperation;
+import com.mongodb.operation.MapReduceToCollectionOperation;
+import com.mongodb.operation.MapReduceWithInlineResultsOperation;
+import com.mongodb.operation.MixedBulkWriteOperation;
+import com.mongodb.operation.RenameCollectionOperation;
+import org.bson.BsonDocument;
+import org.bson.BsonDocumentWrapper;
+import org.bson.BsonJavaScript;
+import org.bson.BsonString;
+import org.bson.codecs.Codec;
+import org.bson.codecs.CollectibleCodec;
+import org.bson.codecs.configuration.CodecRegistry;
+import org.bson.conversions.Bson;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+import static com.mongodb.assertions.Assertions.notNull;
+import static java.lang.String.format;
+import static java.util.Collections.singletonList;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+final class Operations<TDocument> {
+    private final MongoNamespace namespace;
+    private final Class<TDocument> documentClass;
+    private final ReadPreference readPreference;
+    private final CodecRegistry codecRegistry;
+    private final WriteConcern writeConcern;
+    private final boolean retryWrites;
+    private final ReadConcern readConcern;
+
+    Operations(final MongoNamespace namespace, final Class<TDocument> documentClass, final ReadPreference readPreference,
+                      final CodecRegistry codecRegistry, final WriteConcern writeConcern, final boolean retryWrites,
```

Comment:
Comment: our styles have gone out of sync again, my settings would have these inline

Reply:
Nah, I just neglected to reformat after refactoring.  Will fix.

Reply:
Done